File: CSharp/Medium/Backtracking/GenerateParantheses.cs
/*
For example, given n = 3, a solution set is:

[
  "((()))",
  "(()())",
  "(())()",
  "()(())",
  "()()()"
]
*/
public class Solution {
    public IList<string> GenerateParenthesis(int n) {
        IList<string> result = new List<string>();
        string resultStr = "";
        if(n==0)
        {
        	return result;
        }
        GenerateParantheses(0,0,result,resultStr,n);
        return result;
    }

    private void GenerateParantheses(int openParantheses, int closeParantheses,IList<string> result, string resultStr, int n)
    {
    	if(openParantheses==closeParantheses && openParantheses+closeParantheses == 2*n)
    	{
    		result.Add(resultStr);
    	}
    	else
    	{
    		if(openParantheses<n)
	    	{
	    		GenerateParantheses(openParantheses+1,closeParantheses,result,resultStr+"(",n);
	    	}
	    	if(closeParantheses<openParantheses)
	    	{
	    		GenerateParantheses(openParantheses,closeParantheses+1,result,resultStr+")",n);
	    	}
    	} 
    }
}

File: Java/Moderate/Backtracking/GenerateParantheses.java
/*
For example, given n = 3, a solution set is:

[
  "((()))",
  "(()())",
  "(())()",
  "()(())",
  "()()()"
]*/
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> result = new ArrayList<String>();
        if(n==0)
        {
        	return result;
        }
        generateParenthesis(n,0,0,result,"");
        return result;
    }
    private void generateParenthesis(int n, int openCount, int closeCount, List<String> result, String str)
    {
    	if(openCount+closeCount==n*2)
    	{
    		result.add(str);
    	}
    	else
    	{
    		if(openCount<n)
	    	{
	    		generateParenthesis(n,openCount+1,closeCount,result,str+"(");
	    	}
	    	if(closeCount<openCount)
	    	{
	    		generateParenthesis(n,openCount,closeCount+1,result,str+")");
	    	}
    	}
    }
}